### PR TITLE
Fix Python support in integration test framework

### PR DESCRIPTION
**Pull request by swgillespie**
Update the integratino test framework to use pipenv to bootstrap new
virtual environments for tests and use those virtual environments to run
pulumi update and pulumi preview.

Fixes pulumi/pulumi#2138

### CODE REVIEW

File: pkg/testing/integration/program.go [hunk 1]
```diff
@@ -1107,7 +1181,31 @@ func (pt *programTester) prepareNodeJSProject(projinfo *engine.Projinfo) error {
...
+		return err
+	}
+
+	// Install the package's dependencies, which Pipenv infers from a requirements.txt file in the root of the project.
+	if err = pt.runPipenvCommand("pipenv-install", []string{"install"}, cwd); err != nil {
+		return err
+	}
+
+	// Install the Python SDK that we are testing into the virtualenv. This is the SDK that we just built.
+	sdkPath, err := pt.getPythonSDKPath()
```
Comment (ellismg) [inline pkg/testing/integration/program.go]: I am not sure this is going to be correct in up-stack repositories.  In these repositories, we'd want to install the runtime from PyPI but the package under test in this way, correct?

I think that instead we should probably be using the the `Dependencies` array on ProgramTestOptions in some way (maybe it is a path relative to `GOPATH/src`?), similar to how we use these as a way to signal what to `yarn link`?

Comment (swgillespie) [reply pkg/testing/integration/program.go]: Ah, you're right. I spend too much time in pulumi/pulumi.

Unfortunately pipenv doesn't have any kind of `yarn link` equivalent, so for each repo we're going to need to tell the integration test where to look for the package to install. I think I can get this worked out, though.

**Comment by swgillespie** [comment]
@ellismg fixed in the most recent commit. The strategy I tried here is to install all packages that don't contain OS path separators as if they are PyPI packages, while installing paths as if they were references to the filesystem. Upstream, this lets us refer to packages by their bare names if we want to install them through PyPI (e.g. `pulumi`) while being able to use relative paths to test packages we have just built.

**Comment by ellismg** [comment]
> The strategy I tried here is to install all packages that don't contain OS path separators as if they are PyPI packages, while installing paths as if they were references to the filesystem.

Fine with this approach, but FWIW, this differs from how we do things for node.  In node, we still have a package.json that depends on every package that isn't under test, and run `yarn install` to pull in these dependencies as part of the setup.  The entries in `Dependenices` get `yarn link`'ed unconditionally.

So I would have imagined that we use `requirements.txt` here for every dependent package except the one ones under test, for which we do the `pipenv install -e` thing.  I have a slight preference for trying to use `requirements.txt` as much as possible, but if you think this is better, I'm cool with it.

**Comment by ellismg** [comment]
> I have a slight preference for trying to use requirements.txt as much as possible, but if you think this is better, I'm cool with it.

Also happy if this is just more expedient for now and we backlog thinking about our long term plans for Python end-to-end testing.

File: pkg/testing/integration/program.go [hunk 2]
```diff
@@ -1107,6 +1166,49 @@ func (pt *programTester) prepareNodeJSProject(projinfo *engine.Projinfo) error {
 
 // preparePythonProject runs setup necessary to get a Python project ready for `pulumi` commands.
 func (pt *programTester) preparePythonProject(projinfo *engine.Projinfo) error {
+	cwd, _, err := projinfo.GetPwdMain()
+	if err != nil {
+		return err
+	}
+
+	// Create a new Pipenv environment. This bootstraps a new virtual environment containing the version of Python that
+	// we requested. Note that this version of Python is sourced from the machine, so you must first install the version
```
Comment (ellismg) [inline pkg/testing/integration/program.go]: We allow the version of Python to be configured per-test.  Should we?  Is there a well known way to install multiple Python 3.X versions concurrently?

Comment (swgillespie) [reply pkg/testing/integration/program.go]: > We allow the version of Python to be configured per-test. Should we? Is there a well known way to install multiple Python 3.X versions concurrently?

Yeah, it's possible and not uncommon to do this. There's even a [PEP](https://www.python.org/dev/peps/pep-0394/) to document a convention for having different versions of Python on the path. In general there's nothing stopping people from having multiple `python3.x` executables on their path pointing to different Python versions. It's certainly possible to do so; the Python toolchain has evolved around the necessity to install Python 2 and 3 side-by-side so it's actually not to hard to do this today with Python 3 versions.

As to whether or not we should let tests select their version of Python, I doubt it's going to be useful in the near term. I don't mind removing it.

**Comment by swgillespie** [comment]
> I have a slight preference for trying to use requirements.txt as much as possible, but if you think this is better, I'm cool with it.

That's fine with me . I'll go ahead and do that now.

**Comment by swgillespie** [comment]
Actually, to be clear, this setup will already work today in the presence of `requirements.txt`. `pipenv --python 3.6` will read `requirements.txt` to populate its package constraints and install them as part of `pipenv install`. So really, even as the code is written now, you can accomplish what you're describing, but you can *also* depend on packages explicitly via `Dependencies` if you need to.

Given that we don't need to install packages via Dependencies if we use requirements.txt, I'll just delete that code.

**Comment by swgillespie** [comment]
thanks!

**Pull request merged**